File: environment/src/main/java/jetbrains/exodus/tree/patricia/NodeBase.java
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.tree.patricia;

import jetbrains.exodus.ArrayByteIterable;
import jetbrains.exodus.ByteIterable;
import jetbrains.exodus.ByteIterator;
import jetbrains.exodus.log.ByteIterableWithAddress;
import jetbrains.exodus.log.ByteIteratorWithAddress;
import jetbrains.exodus.log.CompressedUnsignedLongByteIterable;
import jetbrains.exodus.tree.INode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.PrintStream;

@SuppressWarnings({"ProtectedField"})
abstract class NodeBase implements INode {

    private static final int LAZY_KEY_VALUE_ITERABLE_MIN_LENGTH = 16;

    @NotNull
    protected ByteIterable keySequence;
    @Nullable
    protected ByteIterable value;

    NodeBase(@NotNull final ByteIterable keySequence, @Nullable final ByteIterable value) {
        this.keySequence = keySequence;
        this.value = value;
    }

    NodeBase(final byte type,
             @NotNull final ByteIterableWithAddress data,
             @NotNull final ByteIteratorWithAddress it) {
        this.keySequence = extractKey(type, data, it);
        this.value = extractValue(type, data, it);
    }

    long matchesKeySequence(@NotNull final ByteIterator it) {
        int matchingLength = 0;
        final ByteIterator keyIt = keySequence.iterator();
        while (keyIt.hasNext()) {
            final byte keyByte = keyIt.next();
            if (!it.hasNext()) {
                return MatchResult.getMatchResult(-matchingLength - 1, keyByte, false, (byte) 0);
            }
            final byte nextByte = it.next();
            if (nextByte != keyByte) {
                return MatchResult.getMatchResult(-matchingLength - 1, keyByte, true, nextByte);
            }
            ++matchingLength;
        }
        return MatchResult.getMatchResult(matchingLength);
    }

    boolean hasKey() {
        return keySequence != ByteIterable.EMPTY && keySequence.getLength() > 0;
    }

    @NotNull
    @Override
    public ByteIterable getKey() {
        return keySequence;
    }

    @Override
    public boolean hasValue() {
        return value != null;
    }

    @Override
    @Nullable
    public ByteIterable getValue() {
        return value;
    }

    @Override
    public void dump(PrintStream out, int level, @Nullable ToString renderer) {
        throw new UnsupportedOperationException();
    }

    abstract long getAddress();

    abstract boolean isMutable();

    abstract MutableNode getMutableCopy(@NotNull final PatriciaTreeMutable mutableTree);

    @Nullable
    abstract NodeBase getChild(@NotNull final PatriciaTreeBase tree, final byte b);

    @NotNull
    abstract NodeChildrenIterator getChildren(final byte b);

    @NotNull
    abstract NodeChildrenIterator getChildrenRange(final byte b);

    @NotNull
    abstract NodeChildrenIterator getChildrenLast();

    @NotNull
    abstract NodeChildren getChildren();

    abstract int getChildrenCount();

    @Override
    public String toString() {
        return String.format("%s} %s %s",
            keySequence.iterator().hasNext() ? "{key:" + keySequence.toString() : '{',
            value == null ? "@" : value.toString() + " @", getAddress()
        );
    }

    static void indent(PrintStream out, int level) {
        for (int i = 0; i < level; i++) {
            out.print(' ');
        }
    }

    static class MatchResult {

        static long getMatchResult(final int matchingLength) {
            return getMatchResult(matchingLength, (byte) 0, false, (byte) 0);
        }

        static long getMatchResult(final int matchingLength,
                                   final byte keyByte,
                                   final boolean hasNext,
                                   final byte nextByte) {
            long result = (((long) Math.abs(matchingLength)) << 18) + ((keyByte & 0xff) << 10) + ((nextByte & 0xff) << 2);
            if (matchingLength < 0) {
                result += 2;
            }
            if (hasNext) {
                ++result;
            }
            return result;
        }

        static int getMatchingLength(final long matchResult) {
            final int result = (int) (matchResult >> 18);
            return (matchResult & 2) == 0 ? result : -result;
        }

        static int getKeyByte(final long matchResult) {
            return (int) (matchResult >> 10) & 0xff;
        }

        static int getNextByte(final long matchResult) {
            return (int) (matchResult >> 2) & 0xff;
        }

        static boolean hasNext(final long matchResult) {
            return (matchResult & 1) != 0;
        }
    }

    final class EmptyNodeChildrenIterator implements NodeChildrenIterator {

        @Override
        public boolean isMutable() {
            return false;
        }

        @Override
        public void nextInPlace() {
            throw new UnsupportedOperationException();
        }

        @Override
        public void prevInPlace() {
            throw new UnsupportedOperationException();
        }

        @Override
        public ChildReference getNode() {
            return null;
        }

        @Override
        public NodeBase getParentNode() {
            return NodeBase.this;
        }

        @Override
        public int getIndex() {
            return 0;
        }

        @Override
        public ByteIterable getKey() {
            return ByteIterable.EMPTY;
        }

        @Override
        public boolean hasNext() {
            return false;
        }

        @Override
        public ChildReference next() {
            return null;
        }

        @Override
        public boolean hasPrev() {
            return false;
        }

        @Override
        public ChildReference prev() {
            return null;
        }

        @Override
        public void remove() {
        }
    }

    @NotNull
    private static ByteIterable extractKey(final byte type,
                                           @NotNull final ByteIterableWithAddress data,
                                           @NotNull final ByteIteratorWithAddress it) {
        if (!PatriciaTreeBase.nodeHasKey(type)) {
            return ByteIterable.EMPTY;
        }
        return extractLazyIterable(data, it);
    }

    @Nullable
    private static ByteIterable extractValue(final byte type,
                                             @NotNull final ByteIterableWithAddress data,
                                             @NotNull final ByteIteratorWithAddress it) {
        if (!PatriciaTreeBase.nodeHasValue(type)) {
            return null;
        }
        return extractLazyIterable(data, it);
    }

    private static ByteIterable extractLazyIterable(@NotNull final ByteIterableWithAddress data,
                                                    @NotNull final ByteIteratorWithAddress it) {
        final int length = CompressedUnsignedLongByteIterable.getInt(it);
        if (length == 1) {
            return ArrayByteIterable.fromByte(it.next());
        }
        if (length < LAZY_KEY_VALUE_ITERABLE_MIN_LENGTH) {
            return new ArrayByteIterable(it, length);
        }
        final ByteIterable result = data.subIterable((int) (it.getAddress() - data.getDataAddress()), length);
        it.skip(length);
        return result;
    }
}

File: environment/src/main/kotlin/jetbrains/exodus/tree/patricia/NodeBase.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.tree.patricia

import jetbrains.exodus.ArrayByteIterable
import jetbrains.exodus.ByteIterable
import jetbrains.exodus.ByteIterator
import jetbrains.exodus.log.ByteIterableWithAddress
import jetbrains.exodus.log.ByteIteratorWithAddress
import jetbrains.exodus.tree.Dumpable
import jetbrains.exodus.tree.INode
import java.io.PrintStream
import java.util.NoSuchElementException
import kotlin.math.abs

internal abstract class NodeBase : INode {
    @JvmField
    internal var key: ByteIterable

    @JvmField
    internal var value: ByteIterable?

    constructor(keySequence: ByteIterable, value: ByteIterable?) {
        key = keySequence
        this.value = value
    }

    constructor(
        type: Byte,
        data: ByteIterableWithAddress,
        it: ByteIteratorWithAddress
    ) {
        key = extractKey(type, data, it)
        value = extractValue(type, data, it)
    }

    override fun getKey(): ByteIterable = key

    override fun getValue(): ByteIterable? = value

    fun matchesKeySequence(it: ByteIterator): Long {
        var matchingLength = 0
        val keyIt = key.iterator()
        if (keyIt is ArrayByteIterable.Iterator && it is ArrayByteIterable.Iterator) {
            matchingLength = keyIt.match(it)
            if (!keyIt.hasNext()) {
                return MatchResult.getMatchResult(matchingLength)
            }
            val keyByte = keyIt.next()
            if (!it.hasNext()) {
                return MatchResult.getMatchResult(-matchingLength - 1, keyByte, false, 0.toByte())
            }
            val nextByte = it.next()
            return MatchResult.getMatchResult(-matchingLength - 1, keyByte, true, nextByte)
        }
        while (keyIt.hasNext()) {
            val keyByte = keyIt.next()
            if (!it.hasNext()) {
                return MatchResult.getMatchResult(-matchingLength - 1, keyByte, false, 0.toByte())
            }
            val nextByte = it.next()
            if (nextByte != keyByte) {
                return MatchResult.getMatchResult(-matchingLength - 1, keyByte, true, nextByte)
            }
            ++matchingLength
        }
        return MatchResult.getMatchResult(matchingLength)
    }

    fun hasKey(): Boolean {
        return key !== ByteIterable.EMPTY && key.length > 0
    }

    override fun hasValue(): Boolean {
        return value != null
    }

    override fun dump(out: PrintStream, level: Int, renderer: Dumpable.ToString?) {
        throw UnsupportedOperationException()
    }

    abstract fun getAddress(): Long
    abstract fun isMutable(): Boolean
    abstract fun getMutableCopy(mutableTree: PatriciaTreeMutable): MutableNode
    abstract fun getChild(tree: PatriciaTreeBase, b: Byte): NodeBase?
    abstract fun getChildren(b: Byte): NodeChildrenIterator
    abstract fun getChildrenRange(b: Byte): NodeChildrenIterator
    abstract fun getChildrenLast(): NodeChildrenIterator
    abstract fun getChildren(): NodeChildren
    abstract fun getChildrenCount(): Int
    override fun toString(): String {
        return String.format(
            "%s} %s %s",
            if (key.iterator().hasNext()) "{key:$key" else '{',
            if (value == null) "@" else value.toString() + " @", getAddress()
        )
    }

    internal object MatchResult {
        fun getMatchResult(matchingLength: Int): Long {
            return getMatchResult(matchingLength, 0.toByte(), false, 0.toByte())
        }

        fun getMatchResult(
            matchingLength: Int,
            keyByte: Byte,
            hasNext: Boolean,
            nextByte: Byte
        ): Long {
            var result = (abs(matchingLength)
                .toLong() shl 18) + (keyByte.toInt() and 0xff shl 10) + (nextByte.toInt() and 0xff shl 2)
            if (matchingLength < 0) {
                result += 2
            }
            if (hasNext) {
                ++result
            }
            return result
        }

        fun getMatchingLength(matchResult: Long): Int {
            val result = (matchResult shr 18).toInt()
            return if (matchResult and 2L == 0L) result else -result
        }

        fun getKeyByte(matchResult: Long): Int {
            return (matchResult shr 10).toInt() and 0xff
        }

        fun getNextByte(matchResult: Long): Int {
            return (matchResult shr 2).toInt() and 0xff
        }

        fun hasNext(matchResult: Long): Boolean {
            return matchResult and 1L != 0L
        }
    }

    internal inner class EmptyNodeChildrenIterator : NodeChildrenIterator {
        override fun getKey(): ByteIterable? = ByteIterable.EMPTY
        override fun isMutable(): Boolean = false

        override fun nextInPlace() {
            throw UnsupportedOperationException()
        }

        override fun prevInPlace() {
            throw UnsupportedOperationException()
        }

        override fun getNode(): ChildReference? = null
        override fun getParentNode(): NodeBase = this@NodeBase
        override fun getIndex(): Int = 0

        override fun hasNext(): Boolean {
            return false
        }

        override fun next(): ChildReference {
            throw NoSuchElementException()
        }

        override fun hasPrev(): Boolean {
            return false
        }

        override fun prev(): ChildReference? {
            return null
        }

        override fun remove() {}
    }

    companion object {
        fun indent(out: PrintStream, level: Int) {
            for (i in 0 until level) {
                out.print(' ')
            }
        }

        private fun extractKey(
            type: Byte,
            data: ByteIterableWithAddress,
            it: ByteIteratorWithAddress
        ): ByteIterable {
            return if (!PatriciaTreeBase.nodeHasKey(type)) {
                ByteIterable.EMPTY
            } else extractLazyIterable(data, it)
        }

        private fun extractValue(
            type: Byte,
            data: ByteIterableWithAddress,
            it: ByteIteratorWithAddress
        ): ByteIterable? {
            return if (!PatriciaTreeBase.nodeHasValue(type)) {
                null
            } else extractLazyIterable(data, it)
        }

        private fun extractLazyIterable(
            data: ByteIterableWithAddress,
            it: ByteIteratorWithAddress
        ): ByteIterable {
            val length = it.getCompressedUnsignedInt()
            if (length == 1) {
                return ArrayByteIterable.fromByte(it.next())
            }
            val result = data.cloneWithAddressAndLength(it.getAddress(), length)
            it.skip(length.toLong())
            return result
        }
    }
}